lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2556,15 +2558,29 @@ static void inferDynamic(ASTContext &ctx, ValueDecl *D) {
 
   // Variables declared with 'let' cannot be 'dynamic'.
   if (auto VD = dyn_cast<VarDecl>(D)) {
-    if (VD->isLet() && !isNSManaged) return;
+    auto staticSpelling = VD->getParentPatternBinding()->getStaticSpelling();
+
+    // The presence of 'static' bocks the inference of 'dynamic'.
+    if (VD->isStatic() && staticSpelling == StaticSpellingKind::KeywordStatic)
```

Comment:
Do you need to check `isStatic` if you're already checking the static spelling kind?